static-analysis/teammates-macker.xml

```diff
@@ -1,45 +1,401 @@
+            <ruleset name="TestDriverComponent: Only certain test cases can test storage:entity">
+                <access-rule>
+                    <deny>
+                        <from class="${test}.**" />
+                        <to class="${storage}.entity.**" />
+                        <allow>
+                            <from>
+                                <include class="${testCases}.datatransfer.**" />
+                            </from>
+                        </allow>
+                        
+                        <!-- Violations that need fixing -->
+                        <allow>
+                            <from class="${testCases}.storage.FeedbackResponseCommentsDbTest" />
+                            <to class="${storage}.entity.FeedbackResponseComment" />
+                        </allow>
+                        <allow>
+                            <from class="${testCases}.logic.StudentsLogicTest" />
+                            <to class="${storage}.entity.CourseStudent" />
+                        </allow>
+                    </deny>
+                </access-rule>
+            </ruleset>
+            <ruleset name="TestDriverComponent: Only certain test cases can access GaeSimulation/BackDoorLogic">
+                <access-rule>
+                    <deny>
+                        <from class="${test}.**" />
+                        <to>
+                            <include class="${test}.driver.GaeSimulation" />
+                            <include class="${logic}.backdoor.BackDoorLogic" />
+                        </to>
+                        <allow>
+                            <from>
+                                <include class="${testCases}.storage.**" />
+                                <include class="${testCases}.logic.**" />
+                                <include class="${testCases}.action.**" />
+                                <include class="${testCases}.automated.**" />
+                                <include class="${testCases}.search.**" />
+                                <include class="${testCases}.BaseComponentTestCase" />
+                                
+                                <!-- Violation that needs fixing -->
+                                <include class="${testCases}.pagedata.StudentFeedbackResultsPageDataTest" />
+                            </from>
+                        </allow>
+                    </deny>
+                </access-rule>
+            </ruleset>
+            <ruleset name="TestDriverComponent: BackDoor test driver is only for browser test">
+                <access-rule>
+                    <deny>
+                        <to class="${test}.driver.BackDoor" />
+                        <allow>
+                            <from>
+                                <include class="${testCases}.browsertests.**" />
+                                <include class="${test}.driver.BackDoor**$**" /> <!-- allow dependency to inner class -->
```

Comment:
To make it more organized, put this one before the `${testCases}.browsertests.**` entry.